### DISCUSSION TITLE
Questions on Enable-D365MaintenanceMode

### DISCUSSION THREAD

**Discussion by FH-Inway** [d365collaborative/d365fo.tools#585] [Q&A] [answered]
Today I had the rare opportunity to enable the maintenance mode on a cloud hosted environment (demo topology). I was thinking about using the [Enable-D365MaintenanceMode](https://github.com/d365collaborative/d365fo.tools/blob/c34e0fd70311f7b368c618f0a54c29a917428d53/d365fo.tools/functions/enable-d365maintenancemode.ps1) cmdlet, but ended up doing it the [documented way](https://docs.microsoft.com/en-us/dynamics365/fin-ops-core/dev-itpro/sysadmin/maintenance-mode#turn-maintenance-mode-on-and-off-in-devtestdemo-environments-hosted-in-customers-subscription).

The cmdlets documentation states that the Microsoft.Dynamics.AX.Deployment.Setup.exe is used to enable the maintenance mode. Since the documentation suggests this should only be done on VHD-based environments, I'm wondering if using Microsoft.Dynamics.AX.Deployment.Setup.exe on a CHE is fine, too.

I also noticed that the cmdlet references the ```disable-maintenancemode.sql``` script. I haven't got a chance to test this, but it looks like a copy&paste error. I think the ```enable-maintenancemode.sql``` script should be referenced instead.

**Comment by Splaxi** [comment, marked as answer]
You are totally right  !

We should refactor both cmdlet to offer a mandatory parameter where you can select one onebox or che mode.

**Comment by FH-Inway** [comment]
Sounds good, I will give it a shot. Are there other cmdlets that have something similar? Or maybe some way of detecting if the tools are running on a CHE or VHD environment? This way, a mandatory parameter could be avoided.

**Comment by Splaxi** [comment]
Not really. 

Back in the days we tested for elevated privileges - that indicated one onebox. 

But thinking about it, the sql password could be an indication we could trust. The same goes for the package directory - if on  c drive, it will most likely be a onebox 🤞

**Comment by Splaxi** [comment]
Or the certificates on the iis might be even better 🤘

**Comment by FH-Inway** [comment]
What about the ```$Script:IsOnebox``` variable? Several cmdlets make use of it and the tools use one of the application dlls to set it (see [Get-ApplicationEnvironment](https://github.com/d365collaborative/d365fo.tools/blob/a60306e1a478f802891d0a32091d0826d3b2b837/d365fo.tools/internal/functions/get-applicationenvironment.ps1))

**Comment by Splaxi** [comment]
IsOneboxEnvironment is the internal property that we depend on for that variable. 

But if I remember correctly, both che and onebox (vhd) thinks of themselves as IsOneboxEnvironment = $true

**Comment by FH-Inway** [comment]
Will test this. If it is true for both environment types, there is also ```$Script:EnvironmentType``` which gets set in [variables.ps1](https://github.com/d365collaborative/d365fo.tools/blob/69f276e9e476ac14b011148ecdd8ebc1141faace/d365fo.tools/internal/scripts/variables.ps1) based on the URL of the environment.

**Comment by Splaxi** [comment]
True and I would trust that one even more after looking at the code

**Comment by FH-Inway** [comment]
Seems we have a winner. I created issue #586 for further work on this.
I think I also found another issue with the enable script. It starts by stopping all services, but in the end, only the Aos service is started again.

**Reply by Splaxi** [reply]
That is on purpose, as you really only want the aos to be running while in maintenance mode. 

😉

**Reply by FH-Inway** [reply]
Ah, yes, of course, that makes sense.
